### PR TITLE
test(report): per-element tests of coverage data and coverage reports

### PR DESCRIPTION
**Pull request by galtm**
This pull request establishes a baseline for end-to-end testing of XSLT code coverage reporting at a per-element level for most XSLT elements. Adrian Bird (@birdya22 in GitHub) did the majority of the design and development work; see 87f9ea6e7afab61110f3bb4275ac47d3f683591f and 74c6f4789d4a26852cbc2f434cb9d8b2357de8ed.

While reviewing, I added some additional test cases here and there and made minor tweaks.

### Notes

* Adrian, thank you so much for getting this effort started in such a systematic and well-thought-out way! Any of the changes that I made here are open for your review. There are a lot of files, but many of them are either generated or your originals. Consider using the Commits tab to look at changes in groups instead of scrolling through the file list.

* @AirQuick, I'm adding you as a code reviewer as well, although you haven't been active in this repo for several months. Your input is welcome, even if it's after this PR is merged.

* Some of the coverage reports in the `cases-coverage/expected/` directory are wrong due to bugs in either Saxon 12.4 or XSpec. Whenever we're able to resolve those issues in a subsequent pull request, the updates in the expected results will show the effect of the bug fixes.

* Some of the coverage reports in the `cases-coverage/expected/` directory might not reflect the desired design. Consider the files to be a baseline for future comparison if/when we change XSpec code for coverage reporting.

* In addition, `xsl-on-non-empty-01.xspec` has a test failure in Saxon-HE 12.4J that may be related to https://saxonica.plan.io/issues/6428 . With Saxon 12.3J, I also get different coverage XML and HTML, between the HE and EE editions. I'm not able to run Saxon-EE 12.4J, so I can't compare it with Saxon-HE 12.4J. The expected results in this pull request are from Saxon-HE 12.4J, which is also the version that runs in GitHub.

* Soon after this PR is complete, I would like to complete #1918. After that, we can work on other bug fixes and possible design changes, with these tests to help us understand effects and protect against regressions.

### CODE REVIEW

File: test/end-to-end/cases-coverage/xsl-evaluate-01.xsl
Comment (birdya22) [inline test/end-to-end/cases-coverage/xsl-evaluate-01.xsl:1]: Minor comment - remove the "need a test case using with-param" on line 4.

File: test/end-to-end/cases-coverage/xsl-map-01.xsl
Comment (birdya22) [inline test/end-to-end/cases-coverage/xsl-map-01.xsl:1]: The comment on line 6 should end in "includes map-entry" and not "includes map".

File: test/end-to-end/cases-coverage/xsl-map-01.xspec
Comment (birdya22) [inline test/end-to-end/cases-coverage/xsl-map-01.xspec:1]: In the scenario label can you add "(including xsl:map-entry)"

**Review by birdya22 — COMMENTED** [review]
Completed review of the commits. 3 Minor comments on individual files.

> RE: With Saxon 12.3J, I also get different coverage XML and HTML, between the HE and EE editions. 

My guess is that the default optimizations in HE and EE are producing the differences. Is it easy to re-run the tests with all optimizations disabled just to check?

File: test/end-to-end/cases-coverage/xsl-evaluate-01.xsl
Comment (galtm) [reply test/end-to-end/cases-coverage/xsl-evaluate-01.xsl:1]: Done in 23b3c17d8896793a28584963b56736bc7872999e

File: test/end-to-end/cases-coverage/xsl-map-01.xspec
Comment (galtm) [reply test/end-to-end/cases-coverage/xsl-map-01.xspec:1]: Done in 23b3c17d8896793a28584963b56736bc7872999e

**Comment by galtm** [comment]
> Completed review of the commits. 3 Minor comments on individual files.
> 
> > RE: With Saxon 12.3J, I also get different coverage XML and HTML, between the HE and EE editions.
> 
> My guess is that the default optimizations in HE and EE are producing the differences. Is it easy to re-run the tests with all optimizations disabled just to check?

Thanks for your review! Later today, I'll see if I can re-run the tests with all optimizations disabled. I'll keep you posted.

**Comment by galtm** [comment]
> > > RE: With Saxon 12.3J, I also get different coverage XML and HTML, between the HE and EE editions.
> > 
> > 
> > My guess is that the default optimizations in HE and EE are producing the differences. Is it easy to re-run the tests with all optimizations disabled just to check?
> 
> Thanks for your review! Later today, I'll see if I can re-run the tests with all optimizations disabled. I'll keep you posted.

Passing the parameter `-opt:0` doesn't change these results for `xsl-on-non-empty-01.xspec`, assuming the parameter is taking effect. (I'm adding it in the `saxon.custom.options` field on the Parameters tab of the Edit Ant Scenario dialog in Oxygen. That field does work the way I expect for other Saxon options, such as `-T` and `-ea`.)

The coverage XML file has two lines that are different between HE and EE, and the 12.3 HE data matches the 12.4 HE data. The differences are only in the `columnNumber` values.

### Saxon-EE data
```xml
   <traceable traceableId="3" class="net.sf.saxon.expr.instruct.ConditionalBlock"/>
...
   <traceable traceableId="5" class="net.sf.saxon.expr.instruct.OnNonEmptyExpr"/>
   <hit lineNumber="11" columnNumber="43" moduleId="0" traceableId="5"/>
...
   <hit lineNumber="28" columnNumber="57" moduleId="0" traceableId="3"/>
```

### Saxon-HE data
```xml
   <traceable traceableId="3" class="net.sf.saxon.expr.instruct.ConditionalBlock"/>
...
   <traceable traceableId="5" class="net.sf.saxon.expr.instruct.OnNonEmptyExpr"/>
   <hit lineNumber="11" columnNumber="0" moduleId="0" traceableId="5"/>
...
   <hit lineNumber="28" columnNumber="22" moduleId="0" traceableId="3"/>
```

### XSLT code

In the XSLT, line 11 is
```xml
        <xsl:on-non-empty select="'00'" />
```
while line 28 is
```xml
        <xsl:sequence select="@nonexistent-attribute" />
```

**Comment by birdya22** [comment]
> The coverage XML file has two lines that are different between HE and EE, and the 12.3 HE data matches the 12.4 HE data. The differences are only in the columnNumber values.

The EE values are good but I've got no idea why they differ from HE.

**Comment by galtm** [comment]
> > The coverage XML file has two lines that are different between HE and EE, and the 12.3 HE data matches the 12.4 HE data. The differences are only in the columnNumber values.
> 
> The EE values are good but I've got no idea why they differ from HE.

I agree. I was able to reproduce the line 28 phenomenon outside XSpec and created https://saxonica.plan.io/issues/6448 . I had some trouble reproducing the line 11 phenomenon outside XSpec.

**Pull request merged**